doc/man/babeltrace2-log.1.txt
babeltrace2-log(1)
=================
:manpagetype: program
:revdate: 5 October 2017


NAME
----
babeltrace2-log - Convert a Linux kernel ring buffer to a CTF trace


SYNOPSIS
--------
[verse]
*babeltrace2-log* [opt:--with-timestamps] 'OUTPUT-PATH'


DESCRIPTION
-----------
The `babeltrace2-log` tool reads the lines of a Linux kernel ring buffer,
as printed by the man:dmesg(1) tool, from the standard input stream and
converts them to a http://diamon.org/ctf/[CTF] trace written to the
'OUTPUT-PATH' directory.

Usage example:

[role="term"]
----
$ dmesg | babeltrace2-log --with-timestamps my-trace
----

The events of the generated CTF trace are named `string` and contain a
single payload string field named `str` which contains the corresponding
ring buffer line.

By default, `babeltrace2-log` does not try to extract the timestamps of
the kernel ring buffer lines to use them as the created events's
timestamps. A typical man:dmesg(1) line looks like this:

----
[87166.510937] PM: Finishing wakeup.
----

In the last example, the `[87166.510937]` part is a timestamp which
could be extracted. You can make `babeltrace2-log` extract timestamps
from lines with the opt:--with-timestamps option.


OPTIONS
-------
opt:-t, opt:--with-timestamps::
    Extract timestamps from the kernel ring buffer lines: set
    the created event's payload's `str` field to the rest of the line,
    excluding any timestamp prefix.


ENVIRONMENT VARIABLES
---------------------
See the environment variables of man:babeltrace2-source.text.dmesg(7),
man:babeltrace2-filter.utils.muxer(7), and
man:babeltrace2-sink.ctf.fs(7).

include::common-lib-env.txt[]

include::common-cli-files.txt[]

include::common-cmd-footer.txt[]


SEE ALSO
--------
man:babeltrace2-intro(7),
man:babeltrace2-source.text.dmesg(7),
man:babeltrace2-filter.utils.muxer(7),
man:babeltrace2-sink.ctf.fs(7)


ChangeLog
2019-05-03 babeltrace 2.0.0-pre5 (No Pants Day)
	* Fix: flt.lttng-utils.debug-info: build id note section non-native byte order
	* common: add bt_common_read() function
	* Fix: common: va_list type is implementation dependant
	* Fix: uninitialized variant fc in set_field_refs()
	* src.ctf.fs: make trace-info query accept clock-class-offset-{s,ns} parameters
	* ctf: Use g_time_val_from_iso8601 instead of g_date_time_new_from_iso8601
	* tests: update EXTRA_DIST of Python bindings
	* tests: remove non-existent files from EXTRA_DIST
	* include: add missing files to dist tarball
	* bt2: Add wrapper for bt_plugin_get_version
	* Add git-review config
	* Fix: bt2: enum-conversion warning in native_bt_component_class.i
	* Fix: bt2: int-to-pointer-cast warnings in SWIG utils functions
	* sink.ctf.fs: set_field_refs(): make `fc` `const` as `fc_type` depends on it
	* Fix: sink.ctf.fs: do not keep different `tgt_fc_type` as `tgt_fc` changes
	* lib: bt_graph_{consume,run}(): add unlikely() to unlikely condition
	* src.ctf.fs: compute stream range using entire file info group
	* bt2: update bindings to make test_plugins pass
	* bt2: update bindings to make test_component pass
	* bt2: update bindings to make test_component_class pass
	* lib: Add functions to borrow specialized component classes from specialized components
	* bt2: update value.py, make test_value pass
	* bt2: update object model
	* Fix: string format warnings on 32 bits
	* Fix: unused-function warnings in lib/graph/iterator.c
	* Fix: may-uninitialized warning in trace-ir-metadata-field-class-copy.c
	* Fix: maybe-uninitialized warnings in translate-trace-ir-to-ctf-ir.c
	* Fix: maybe-uninitialized warning in create_relative_field_ref()
	* Fix: unused-variable warnings in lib/graph/iterator.c
	* Fix: maybe-uninitialized warning in bt_ctf_stream_flush()
	* Fix: unused-variable warnings in bt_X_freeze() when not in DEV_MODE
	* Fix: src.ctf.fs: metadata-info: sanitize `path` param
	* Fix: flt.lttng-utils.debug-info: extra colon in `bin` member
	* debug info: Skip test that depends on Python bindings
	* babeltrace: Remove leftovers
	* doc: Make python bindings doc build
	* python-plugin-provider: Make it build
	* Remove clock_class_priority_map.py
	* Fix: headers: make static inline upcasts compatible with C++
	* Fix: headers: add missing end of `extern "C"` curly brackets (for C++)
	* Fix: include/Makefile.am: add missing `babeltrace/ctf-writer/object.h`
	* bt2: Rename files to use singular form
	* bt2: Mass clock_value -> clock_snapshot rename
	* bt2: Mass field_types -> field_class rename
	* bt2: Mass notification -> message rename
	* bt2: Sync native_bt_field_class.i with field-class-const.h
	* lib: have dedicated, unique unsigned and signed enum FC mapping objects
	* tests/bindings/python: Mark all tests as skipped
	* bindings/python/bt2: Make the bt2 package importable
	* lib: Fix function name to match declaration
	* Add env wrapper to facilitate importing the in-tree python bindings
	* bindings/python/bt2: Make the bt2 Python bindings build
	* lib: Make bt_value_null a const pointer
	* ctf: Add dash to stream name suffixes
	* ctf: Use environment fields provided by LTTng to make output trace path
	* cli: put space after comma between array elements
	* Fix: test_convert_args: fix malformed test line
	* test_convert_args: test multiple leftover arguments
	* Fix: test_convert_args: fix expected arguments
	* cli: mimic behavior of BT1 when trying to attach to live session
	* plugins/ctf/fs-src: Session rotation support: merge traces with same uuid
	* cli, plugins/ctf/fs-src: Make src.ctf.fs accept multiple root paths
	* cli: Make append_parameter_to_args accept a bt_value value
	* cli: Remove support for --key and --value arguments
	* cli: Support arrays in parameters
	* cli: Adjust integer range check, replace magic numbers with constants
	* cli: Refactor command line value parsing code
	* cli: Refactor handling of negative numbers
	* Fix: tests/Makefile.am: do not run missing `flt.utils.muxer`
	* Adapt `src.ctf.lttng-live` to current API
	* sink.text.pretty: print inactivity msg handling with BT_LOGD_STR
	* graph: Fix: bt_graph_configure() does not propagate errors
	* Fix: flt.lttng-utils.debug-info: leaking `bin_dir` char array
	* Fix: flt.lttng-utils.debug-info: ".debug" extension not appended
	* lib: field path: have a specific item for "current array element"
	* Fix: bfcr_get_sequence_length_cb(): do not set text array's length
	* Fix: `tests/Makefile.am`: do not run missing tests
	* Fix: `tests/ctf-traces/succeed` traces: change CTF version from 0.1 to to 1.8
	* Fix: `test_packet_seq_num.in`: change expected disc. packets messages order
	* Fix: flt.utils.trimmer: accept inited streams ending without other messages
	* Fix: cli: `flt.utils.trimmer` now expect the `gmt` parameter
	* Remove `test_bt_message_iterator`
	* Fix: cli: Acquire reference on bt_value_null while parsing args
	* flt.lttng-utils.debug-info: cleanup: create dedicated msg_iter destroy function
	* tests: remove unused automake variable
	* Fix: lib: Mismatch in function declarations and definitions
	* flt.lttng-utils.debug-info: Don't copy trace class UUID to output trace class
	* flt.lttng-utils.debug-info: Implement file descriptor cache
	* Cleanup: Move `debug-info` in its own sub-directory
	* Fix: flt.lttng-utils.debug-info: Error in src line reporting
	* flt.lttng-utils.debug-info: cleanup: remove usage old struct name
	* flt.lttng-utils.debug-info: cleanup: only print debug message on error
	* flt.lttng-utils.debug-info: use glib memory and string functions
	* sink.ctf.fs: Remove unused function
	* plugins/ctf: Remove unused uuid-related fields in struct ctf_fs_metadata
	* .editorconfig: add `charset = utf-8`
	* Add .editorconfig file
	* Remove babeltrace 1 Python bindings and tests
	* Remove stale trace-ir test files
	* BT_ASSERT(): update rendering
	* Re-implement BT_ASSERT without using the assert macro
	* Adjust bt_component_borrow_class to match declaration

src/cli/babeltrace2-log.c
/*
 * Copyright 2017 Philippe Proulx <pproulx@efficios.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#include "common/assert.h"
#include <babeltrace2/babeltrace.h>
#include <popt.h>
#include <glib.h>

static
void print_usage(FILE *fp)
{
	fprintf(stderr, "Usage: babeltrace2-log [OPTIONS] OUTPUT-PATH\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Options:\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "  -t, --with-timestamps  Extract timestamps from lines and map them to\n");
	fprintf(stderr, "                         a CTF clock class\n");
}

static
int parse_params(int argc, char *argv[], char **output_path,
		bool *no_extract_ts)
{
	enum {
		OPT_WITH_TIMESTAMPS = 1,
		OPT_HELP = 2,
	};
	static struct poptOption opts[] = {
		{ "with-timestamps", 't', POPT_ARG_NONE, NULL, OPT_WITH_TIMESTAMPS, NULL, NULL },
		{ "help", 'h', POPT_ARG_NONE, NULL, OPT_HELP, NULL, NULL },
		{ NULL, '\0', 0, NULL, 0, NULL, NULL },
	};
	poptContext pc = NULL;
	int opt;
	int ret = 0;
	const char *leftover;

	*no_extract_ts = true;
	pc = poptGetContext(NULL, argc, (const char **) argv, opts, 0);
	if (!pc) {
		fprintf(stderr, "Cannot get popt context\n");
		goto error;
	}

	poptReadDefaultConfig(pc, 0);

	while ((opt = poptGetNextOpt(pc)) > 0) {
		switch (opt) {
		case OPT_HELP:
			print_usage(stdout);
			goto end;
		case OPT_WITH_TIMESTAMPS:
			*no_extract_ts = false;
			break;
		default:
			fprintf(stderr, "Unknown command-line option specified (option code %d)\n",
				opt);
			goto error;
		}
	}

	if (opt < -1) {
		fprintf(stderr, "While parsing command-line options, at option %s: %s\n",
			poptBadOption(pc, 0), poptStrerror(opt));
		goto error;
	}

	leftover = poptGetArg(pc);
	if (!leftover) {
		fprintf(stderr, "Command line error: Missing output path\n");
		print_usage(stderr);
		goto error;
	}

	*output_path = strdup(leftover);
	BT_ASSERT(*output_path);
	goto end;

error:
	ret = 1;

end:
	if (pc) {
		poptFreeContext(pc);
	}

	return ret;
}

int main(int argc, char *argv[])
{
	char *output_path = NULL;
	bool no_extract_ts;
	int retcode;
	GError *error = NULL;
	gchar *bt_argv[] = {
		BT_CLI_PATH,
		"run",
		"--component",
		"dmesg:src.text.dmesg",
		"--params",
		NULL, /* no-extract-timestamp=? placeholder */
		"--component",
		"ctf:sink.ctf.fs",
		"--key",
		"path",
		"--value",
		NULL, /* output path placeholder */
		"--params",
		"single-trace=yes",
		"--connect",
		"dmesg:ctf",
		NULL, /* sentinel */
	};

	retcode = parse_params(argc, argv, &output_path, &no_extract_ts);
	if (retcode) {
		goto end;
	}

	if (no_extract_ts) {
		bt_argv[5] = "no-extract-timestamp=yes";
	} else {
		bt_argv[5] = "no-extract-timestamp=no";
	}

	bt_argv[11] = output_path;
	(void) g_spawn_sync(NULL, bt_argv, NULL,
		G_SPAWN_CHILD_INHERITS_STDIN, NULL, NULL,
		NULL, NULL, &retcode, &error);

	if (error) {
		fprintf(stderr, "Failed to execute \"%s\": %s (%d)\n",
			bt_argv[0], error->message, error->code);
	}

end:
	free(output_path);

	if (error) {
		g_error_free(error);
	}

	return retcode;
}
